src/CustomElements/v1/CustomElements.js

```diff
@@ -0,0 +1,597 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+/**
+ * 2.3
+ * http://w3c.github.io/webcomponents/spec/custom/#dfn-element-definition
+ * @typedef {{
+ *  name: string,
+ *  localName: string,
+ *  constructor: Function,
+ *  connectedCallback: Function,
+ *  disconnectedCallback: Function,
+ *  attributeChangedCallback: Function,
+ *  observedAttributes: Array<string>,
+ * }}
+ */
+var CustomElementDefinition;
+
+(function() {
+  'use strict';
+
+  var doc = document;
+  var win = window;
+
+  // name validation
+  // https://html.spec.whatwg.org/multipage/scripting.html#valid-custom-element-name
+
+  /**
+   * @const
+   * @type {Array<string>}
+   */
+  var reservedTagList = [
+    'annotation-xml',
+    'color-profile',
+    'font-face',
+    'font-face-src',
+    'font-face-uri',
+    'font-face-format',
+    'font-face-name',
+    'missing-glyph',
+  ];
+
+  /** @const */
+  var customNameValidation = /^[a-z][.0-9_a-z]*-[\-.0-9_a-z]*$/;
+  function isValidCustomElementName(name) {
+    return customNameValidation.test(name) && reservedTagList.indexOf(name) === -1;
+  }
+
+  function createTreeWalker(root) {
+    // IE 11 requires the third and fourth arguments be present. If the third
+    // arg is null, it applies the default behaviour. However IE also requires
+    // the fourth argument be present even though the other browsers ignore it.
+    return doc.createTreeWalker(root, NodeFilter.SHOW_ELEMENT, null, false);
+  }
+
+  function isElement(node) {
+    return node.nodeType === Node.ELEMENT_NODE
+  }
+
+  /**
+   * A registry of custom element definitions.
+   *
+   * See https://html.spec.whatwg.org/multipage/scripting.html#customelementsregistry
+   *
+   * @constructor
+   * @property {boolean} polyfilled Whether this registry is polyfilled
+   * @property {boolean} enableFlush Set to true to enable the flush() method
+   *   to work. This should only be done for tests, as it causes a memory leak.
+   */
+  function CustomElementsRegistry() {
+    /** @private {Map<string, CustomElementDefinition>} **/
+    this._definitions = new Map();
+
+    /** @private {Map<Function, CustomElementDefinition>} **/
+    this._constructors = new Map();
+
+    this._whenDefinedMap = new Map();
+
+    /** @private {Set<MutationObserver>} **/
+    this._observers = new Set();
```

Comment:
I'm going to address all the other issues before coming back around on observation vs patching. I'd like to see if we can separate the registry, upgrade and reaction parts from the observation parts and give some choice here.